### PR TITLE
14798 Remaining implementation of Applicant Information

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#14798

*Description of changes:*

Commit 1:
- don't render fee summary until data is loaded
- don't check FF if Jest is running
- added conditional columns to ListPeopleAndRoles
- added Email column to ListPeopleAndRoles
- disable add buttons when there is a valid applicant
- auto-assign Applicant role
- deleted obsolete roles
- used shared Contact Point interface
- added restoration properties to interface
- used sub-component for rule list items
- create RuleListItem component
- fixed applicant information validation
- misc cleanup (including some TODO comments)
- created/updated unit tests
- app version = 4.5.13

Commit 2:
- reordered v-cards
- updated labels
- updated Document Delivery props
- updated DocumentDelivery dependency

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -545,14 +546,14 @@ export default class App extends Vue {
...
+    if (this.iaSurveyId > 0) {
+      if (doNotShow) {
         // save id of survey to hide
         localStorage.setItem('HIDE_SURVEY', this.iaSurveyId)
+      } else {
+        localStorage.removeItem('HIDE_SURVEY')
       }
-    } else {
-      localStorage.removeItem('HIDE_SURVEY')
     }
```
Comment (severinbeauvais) [inline src/App.vue:557]: Cleanup -- should not remove item if we don't have a current survey ID.

File: src/mixins/filing-template-mixin.ts [hunk 2]
```diff
@@ -402,7 +402,9 @@ export default class FilingTemplateMixin extends DateMixin {
           nrNumber: this.getNameRequestNumber
         },
         nameTranslations: this.getNameTranslations,
-        parties: this.orgPersonsToParties(this.getAddPeopleAndRoleStep.orgPeople)
+        parties: this.orgPersonsToParties(this.getAddPeopleAndRoleStep.orgPeople),
+        offices: undefined, // *** TODO: update
+        contactPoint: undefined // *** TODO: update
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:407]: Needed due to required fields in interface.

@ketaki-deodhar please update this.

File: tests/unit/App.spec.ts [hunk 3]
```diff
@@ -908,24 +908,6 @@ describe('Restoration - Restore Business Name for a BEN', () => {
...
-        priorityFees: 0,
-        processingFees: 0,
-        serviceFees: 2.0,
-        tax: {
-          gst: 0,
-          pst: 0
-        },
-        total: 125.0
-      }
-    }))
```
Comment (severinbeauvais) [inline tests/unit/App.spec.ts:927]: Moved (and reduced) inside GET call below.

File: src/App.vue [hunk 4]
```diff
@@ -138,7 +138,8 @@
             </v-col>
 
             <v-col cols="12" lg="3">
-              <aside>
+              <!-- Render fee summary only after data is loaded. -->
+              <aside v-if="haveData">
```
Comment (severinbeauvais) [inline src/App.vue:142]: Still displays (at same time as spinner is hidden):

image

File: src/components/common/RegPeopleAndRoles.vue [hunk 5]
```diff
@@ -163,8 +163,8 @@
           outlined
           color="primary"
           class="btn-outlined-primary mt-6"
-          :disabled="showOrgPersonForm"
-          @click="addOrgPerson(null, PartyTypes.ORGANIZATION)"
+          :disabled="showOrgPersonForm || validApplicantPerson || validApplicantOrg"
```
Comment (severinbeauvais) [inline src/components/common/RegPeopleAndRoles.vue:166]: With an individual:
image

With a business or corporation:
image

With neither:
image

File: src/components/common/RegPeopleAndRoles.vue [hunk 6]
```diff
@@ -18,36 +18,41 @@
...
-            <v-icon v-else>mdi-circle-small</v-icon>
-            <span class="rule-item-txt">{{rule.text}}</span>
-          </li>
+          <RuleListItem class="completing-party-rule"
+            v-if="rule.id === RuleIds.NUM_COMPLETING_PARTY"
+            :key="index"
+            :valid="validNumCompletingParty"
+            :showErrors="getShowErrors"
+            :text="rule.text"
+          />
```
Comment (severinbeauvais) [inline src/components/common/RegPeopleAndRoles.vue:27]: These sub-components might not save much code vertical space, but they're a lot simpler.

File: src/components/common/RegPeopleAndRoles.vue [hunk 7]
```diff
@@ -18,36 +18,41 @@
...
-            <span class="rule-item-txt">{{rule.text}}</span>
-          </li>
+          <RuleListItem class="completing-party-rule pt-2"
+            v-if="rule.id === RuleIds.NUM_COMPLETING_PARTY"
+            :key="index"
+            :valid="validNumCompletingParty"
+            :showErrors="getShowErrors"
+            :text="rule.text"
+          />
+          <RuleListItem class="proprietors-rule pt-2"
```
Comment (severinbeauvais) [inline src/components/common/RegPeopleAndRoles.vue:28]: Note that the spacing class is here in the parent, not in the sub-component. This allows the parent to control the spacing as it sees fit without "inheriting" whitespace from the component. Another way to say this is that the spacing in the component itself is "tight" -- which may be useful if the component is reused on a different page with different spacing requirements.

File: src/components/common/RuleListItem.vue [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<template>
+  <li class="rule-list-item">
```
Comment (severinbeauvais) [inline src/components/common/RuleListItem.vue:2]: In most components, we'd use an `id` here, but there may be several instances of this component on a page, and ids are supposed to be unique, so I'm using a classname here instead.

File: src/mixins/filing-template-mixin.ts [hunk 9]
```diff
@@ -402,7 +400,9 @@ export default class FilingTemplateMixin extends DateMixin {
...
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:405]: @ketaki-deodhar ^^^

File: src/views/Restoration/RestorationReviewConfirm.vue [hunk 10]
```diff
@@ -10,15 +10,21 @@
           </p>
       </header>
 
-      <v-card id="company-summary-vcard" flat class="mt-6">
```
Comment (severinbeauvais) [inline src/views/Restoration/RestorationReviewConfirm.vue:13]: As per uxpin, the Applicant Information comes before the Business Information.

File: src/views/Restoration/RestorationReviewConfirm.vue [hunk 11]
```diff
@@ -27,16 +33,18 @@
...
         <DocumentDelivery
           class="py-8 px-6"
-          :contactValue="getBusinessContact.email"
-          :completingPartyEmail="getUserEmail"
-          contactLabel="Registered Office"
+          contactLabel="Business Office"
+          :contactValue="businessOfficeEmail"
+          additionalLabel="Applicant"
+          :additionalValue="applicantEmail"
+          :showCompletingParty="false"
```
Comment (severinbeauvais) [inline src/views/Restoration/RestorationReviewConfirm.vue:47]: Note that "false" needs to be a bound prop. This doesn't work with `showCompletingParty="false"`.

File: src/views/Restoration/RestorationReviewConfirm.vue [hunk 12]
```diff
@@ -116,6 +126,18 @@ export default class RestorationReviewConfirm extends Vue {
...
+  get businessOfficeEmail (): string {
+    return this.getBusinessContact.email
+  }
+
+  get applicantEmail (): string {
+    const orgPeople = this.getAddPeopleAndRoleStep.orgPeople
+    const applicants = orgPeople.filter(
+      orgPerson => orgPerson.roles.some(role => role.roleType === RoleTypes.APPLICANT)
+    )
+    return applicants[0]?.officer.email
```
Comment (severinbeauvais) [inline src/views/Restoration/RestorationReviewConfirm.vue:138]: If this returns undefined then the DocumentDelivery component displays "Not Entered".

File: src/components/common/RegPeopleAndRoles.vue (hunk 6 above)
Comment (jonathan-longe) [reply src/components/common/RegPeopleAndRoles.vue:27]: That's definitely cleaner!

File: src/components/common/ListPeopleAndRoles.vue [hunk 13]
```diff
@@ -146,19 +154,31 @@ import { PartyTypes, RoleTypes, RouteNames } from '@/enums'
...
   protected activeIndex: number // is NaN for new org/person
 
+  get tableHeaders (): Array<string> {
+    const headers = ['Name', 'Mailing Address']
+    if (this.showDeliveryAddressColumn) headers.push('Delivery Address')
+    if (this.showRolesColumn) headers.push('Roles')
+    if (this.showEmailColumn) headers.push('Email')
+    return headers
+  }
+
```
Comment (jonathan-longe) [inline src/components/common/ListPeopleAndRoles.vue:181]: What do you think about making a sub-component to display a row in the table?  Then, if I'm thinking about this correctly, these properties could be computed row attributes.

**Review by jonathan-longe — APPROVED** [review]
Thanks for the opportunity to review @severinbeauvais

File: src/components/common/ListPeopleAndRoles.vue (hunk 13 above)
Comment (severinbeauvais) [reply src/components/common/ListPeopleAndRoles.vue:181]: I think you're right, but I think this isn't complicated enough, or used elsewhere, that it needs to be its own component IMHO. It's a gray area.

File: src/mixins/filing-template-mixin.ts (hunk 9 above)
Comment (ketaki-deodhar) [reply src/mixins/filing-template-mixin.ts:405]: Got it. I will keep an eye on this while merging :)

**Review by ChrisGabel23 — APPROVED** [review]

**Pull request merged**